File: HisabCore/Sources/HisabCore/HDFCTXTParser.swift
import Foundation

/// Parser for HDFC's delimited .txt statement export — fixed-width columns whose
/// exact extents come from the dash ruler under the header:
///
///     Date      Narration              Chq./Ref.No.   Value Dt  Withdrawal ... Balance
///     --------  ---------------------  -------------  --------  ---------- ... -------
///     03/03/26  UPI-…                  0000119436…    03/03/26      30.00      35,694.53
///               continuation of narration…
///
/// Table semantics (direction, chain verification, refs) are shared with the PDF
/// parser via HDFCStatementTable, so both renditions produce identical identities.
public struct HDFCTXTParser: StatementParser {
    public let source = Source.hdfc

    public init() {}

    private static func decode(_ data: Data) -> String? {
        String(data: data, encoding: .utf8) ?? String(data: data, encoding: .isoLatin1)
    }

    public func canParse(data: Data, filename: String) -> Bool {
        guard !data.starts(with: Array("%PDF".utf8)), !data.starts(with: [0x50, 0x4B]),
              let text = Self.decode(data) else { return false }
        return text.contains("HDFC BANK Ltd.") && text.contains("Statement of accounts")
    }

    public func parse(data: Data, password: String?) throws -> ParsedDocument {
        guard let text = Self.decode(data) else { throw ParseError.unrecognizedFormat }
        let lines = text.components(separatedBy: .newlines)

        guard let ruler = lines.first(where: { Self.isRuler($0) }),
              case let columns = Self.dashRanges(in: ruler), columns.count == 7 else {
            throw ParseError.unrecognizedFormat
        }

        func cell(_ line: [Character], _ index: Int) -> String {
            let range = columns[index]
            guard line.count > range.lowerBound else { return "" }
            // A column may bleed a couple of characters past its ruler on the right.
            let upper = min(line.count, index + 1 < columns.count
                            ? columns[index + 1].lowerBound - 1 : line.count)
            return String(line[range.lowerBound..<max(range.lowerBound, upper)])
                .trimmingCharacters(in: .whitespaces)
        }

        var rows: [HDFCRow] = []
        var open: HDFCRow?

        for rawLine in lines {
            if rawLine.contains("STATEMENT SUMMARY") { break }
            if Self.isFence(rawLine) || Self.isRuler(rawLine) {
                if let finished = open { rows.append(finished); open = nil }
                continue
            }
            let chars = Array(rawLine)
            let date = cell(chars, 0)
            if date.wholeMatch(of: #/\d{2}/\d{2}/\d{2}/#) != nil {
                if let finished = open { rows.append(finished) }
                open = HDFCRow(dateText: date,
                               narration: cell(chars, 1),
                               refText: cell(chars, 2).isEmpty ? nil : cell(chars, 2),
                               withdrawalText: cell(chars, 4).isEmpty ? nil : cell(chars, 4),
                               depositText: cell(chars, 5).isEmpty ? nil : cell(chars, 5),
                               balanceText: cell(chars, 6).isEmpty ? nil : cell(chars, 6))
                continue
            }
            // Narration continuation: date and every non-narration cell blank.
            if open != nil, date.isEmpty {
                let continuation = cell(chars, 1)
                if !continuation.isEmpty,
                   cell(chars, 2).isEmpty, cell(chars, 4).isEmpty,
                   cell(chars, 5).isEmpty, cell(chars, 6).isEmpty {
                    open!.narration += " " + continuation
                }
            }
        }
        if let finished = open { rows.append(finished) }

        return try HDFCStatementTable.parse(rows: rows,
                                            openingBalancePaise: Self.openingBalance(in: text),
                                            period: Self.period(in: text))
    }

    static func isRuler(_ line: String) -> Bool {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        return trimmed.count > 20 && trimmed.allSatisfy { $0 == "-" || $0 == " " }
            && trimmed.contains("--")
    }

    static func dashRanges(in ruler: String) -> [Range<Int>] {
        var ranges: [Range<Int>] = []
        var start: Int?
        for (index, char) in ruler.enumerated() {
            if char == "-" {
                if start == nil { start = index }
            } else if let s = start {
                ranges.append(s..<index)
                start = nil
            }
        }
        if let s = start { ranges.append(s..<ruler.count) }
        return ranges
    }

    static func isFence(_ line: String) -> Bool {
        line.contains("HDFC BANK Ltd") || line.contains("Page No .:")
            || line.contains("Statement of accounts") || line.contains("Statement From")
            || line.contains("Chq./Ref.No.") || line.contains("Account Branch")
            || line.contains("Registered Office") || line.contains("GSTIN")
            || line.contains("End Of Statement")
    }

    static func period(in text: String) -> DatePeriod? {
        guard let match = text.firstMatch(of: #/Statement From\s*:\s*(\d{2}/\d{2}/\d{4})\s+To\s*:\s*(\d{2}/\d{2}/\d{4})/#) else {
            return nil
        }
        let formatter = DateFormatter()
        formatter.calendar = YearMonth.istCalendar
        formatter.timeZone = YearMonth.istCalendar.timeZone
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "dd/MM/yyyy"
        guard let start = formatter.date(from: String(match.1)),
              let end = formatter.date(from: String(match.2)) else { return nil }
        return DatePeriod(start: start, end: end)
    }

    static func openingBalance(in text: String) -> Int64? {
        guard let match = text.firstMatch(of: #/Opening Balance[^\n]*\n\s*([\d,]+\.\d{2})/#) else {
            return nil
        }
        return Money.signedPaise(fromDecimalString: String(match.1))
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/parsers/hdfc_txt_parser.dart
/// HDFC delimited .txt export — fixed-width columns whose extents come from
/// the dash ruler under the header. Port of HDFCTXTParser.swift.
library;

import 'dart:convert';

import '../domain.dart';
import 'hdfc_common.dart';

class HdfcTxtParser implements StatementParser {
  @override
  Source get source => Source.hdfc;

  const HdfcTxtParser();

  static String? _decode(List<int> data) {
    try {
      return utf8.decode(data);
    } on FormatException {
      return latin1.decode(data, allowInvalid: true);
    }
  }

  @override
  bool canParse(List<int> data, String filename) {
    if (data.length >= 4 &&
        String.fromCharCodes(data.take(4)) == '%PDF') {
      return false;
    }
    if (data.length >= 2 && data[0] == 0x50 && data[1] == 0x4B) return false;
    final text = _decode(data);
    if (text == null) return false;
    return text.contains('HDFC BANK Ltd.') &&
        text.contains('Statement of accounts');
  }

  @override
  ParsedDocument parse(List<int> data, {String? password}) {
    final text = _decode(data);
    if (text == null) throw const UnrecognizedFormatException();
    final lines = text.split(RegExp(r'\r?\n'));

    final ruler = lines.where(isRuler).firstOrNull;
    if (ruler == null) throw const UnrecognizedFormatException();
    final columns = dashRanges(ruler);
    if (columns.length != 7) throw const UnrecognizedFormatException();

    String cell(String line, int index) {
      final range = columns[index];
      if (line.length <= range.$1) return '';
      // A column may bleed a couple of characters past its ruler on the right.
      final upper = index + 1 < columns.length
          ? columns[index + 1].$1 - 1
          : line.length;
      final clamped = upper < line.length ? upper : line.length;
      if (clamped <= range.$1) return '';
      return line.substring(range.$1, clamped).trim();
    }

    final rows = <HdfcRow>[];
    HdfcRow? open;
    final dateRegex = RegExp(r'^\d{2}/\d{2}/\d{2}$');

    for (final rawLine in lines) {
      if (rawLine.contains('STATEMENT SUMMARY')) break;
      if (isFence(rawLine) || isRuler(rawLine)) {
        if (open != null) {
          rows.add(open);
          open = null;
        }
        continue;
      }
      final date = cell(rawLine, 0);
      if (dateRegex.hasMatch(date)) {
        if (open != null) rows.add(open);
        String? nonEmpty(String s) => s.isEmpty ? null : s;
        open = HdfcRow(
          dateText: date,
          narration: cell(rawLine, 1),
          refText: nonEmpty(cell(rawLine, 2)),
          withdrawalText: nonEmpty(cell(rawLine, 4)),
          depositText: nonEmpty(cell(rawLine, 5)),
          balanceText: nonEmpty(cell(rawLine, 6)),
        );
        continue;
      }
      // Narration continuation: date and every non-narration cell blank.
      if (open != null && date.isEmpty) {
        final continuation = cell(rawLine, 1);
        if (continuation.isNotEmpty &&
            cell(rawLine, 2).isEmpty &&
            cell(rawLine, 4).isEmpty &&
            cell(rawLine, 5).isEmpty &&
            cell(rawLine, 6).isEmpty) {
          open.narration += ' $continuation';
        }
      }
    }
    if (open != null) rows.add(open);

    return HdfcStatementTable.parse(
      rows: rows,
      openingBalancePaise: HdfcStatementTable.openingBalanceIn(text),
      period: HdfcStatementTable.periodIn(text),
    );
  }

  static bool isRuler(String line) {
    final trimmed = line.trim();
    return trimmed.length > 20 &&
        trimmed.split('').every((c) => c == '-' || c == ' ') &&
        trimmed.contains('--');
  }

  /// Dash runs as (start, end) offsets.
  static List<(int, int)> dashRanges(String ruler) {
    final ranges = <(int, int)>[];
    int? start;
    for (var index = 0; index < ruler.length; index++) {
      if (ruler[index] == '-') {
        start ??= index;
      } else if (start != null) {
        ranges.add((start, index));
        start = null;
      }
    }
    if (start != null) ranges.add((start, ruler.length));
    return ranges;
  }

  static bool isFence(String line) =>
      line.contains('HDFC BANK Ltd') ||
      line.contains('Page No .:') ||
      line.contains('Statement of accounts') ||
      line.contains('Statement From') ||
      line.contains('Chq./Ref.No.') ||
      line.contains('Account Branch') ||
      line.contains('Registered Office') ||
      line.contains('GSTIN') ||
      line.contains('End Of Statement');
}
